pkg/cluster/serviceprincipal.go

```diff
@@ -48,27 +44,21 @@ func (m *manager) clusterSPObjectID(ctx context.Context) error {
 	// wait.PollImmediateUntil. Doing this will not propagate the latest error
 	// to the user in case when wait exceeds the timeout
 	_ = wait.PollImmediateUntil(10*time.Second, func() (bool, error) {
-		var res azgraphrbac.ServicePrincipalObjectResult
-		res, err = m.spApplications.GetServicePrincipalsIDByAppID(ctx, spp.ClientID)
+		clusterSPObjectID, err = utilgraph.GetServicePrincipalIDByAppID(ctx, m.spGraphClient, spp.ClientID)
 		if err != nil {
-			// TODO - with migration to MSGraph, this error is different now.
-			// it's fixed with a refreshing action, but we should fix it here too
-			if strings.Contains(err.Error(), "Authorization_IdentityNotFound") {
-				m.log.Info(err)
-				return false, nil
-			}
 			return false, err
 		}
-
-		clusterSPObjectID = *res.Value
+		if clusterSPObjectID == nil {
+			return false, fmt.Errorf("no service principal found for application ID '%s'", spp.ClientID)
```

Comment:
This might introduce some issues, i'm not 100% sure.  

Since ARM cache is eventually consistent, the creation of an Application may take ~5 minutes to register, which means we should be "waiting" for ARM cache to be eventually consistent (up to 5 minutes) here.  

Our refreshing action won't cover this error case since we're returning a new error.  

Unless MSFT graph doesn't have any cache around creating & immediately using an application + SP.

Reply:
This goes together with my previous comment above.

Reply:
As mentioned on Slack, **azcore** has [built-in logic](https://github.com/Azure/azure-sdk-for-go/blob/b15d1692db75dc448136f239cf9500dd3fffd4b5/sdk/azcore/runtime/policy_bearer_token.go#L88-L95) to handle what our refreshable code is doing (force a token refresh on authentication error).  It does, however, require the client to define its own [AuthorizationHandler](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/azcore@v1.6.1/policy#AuthorizationHandler) callbacks, particularly "OnChallenge".  I see this done for [several azure-sdk-for-go clients](https://github.com/search?q=repo%3AAzure%2Fazure-sdk-for-go+OnChallenge%3A&type=code), but haven't yet confirmed the graph client defines it.